nowcasting_dataset/filesystem/utils.py

```diff
@@ -13,42 +14,58 @@ def upload_and_delete_local_files(dst_path: str, local_path: Path):
     Upload an entire folder and delete local files to either AWS or GCP
     """
     _LOG.info("Uploading!")
-    filesystem = fsspec.open(dst_path).fs
+    filesystem = get_filesystem(dst_path)
     filesystem.put(str(local_path), dst_path, recursive=True)
     delete_all_files_in_temp_path(local_path)
 
 
-def get_maximum_batch_id(path: str):
+def get_filesystem(path: Union[str, Path]) -> fsspec.AbstractFileSystem:
+    r"""Get the fsspect FileSystem from a path.
+
+    For example, if `path` starts with `gs:\\` then return a fsspec.GCSFileSystem.
+
+    It is safe for `path` to include wildcards in the final filename.
+    """
+    path = Pathy(path)
+    return fsspec.open(path.parent).fs
+
+
+# TODO: Issue #308: Use leading zeros in batch filenames, then we can sort the filename strings
+# and take the last one, instead of converting all filenames to ints!
+def get_maximum_batch_id(path: Pathy) -> int:
     """
     Get the last batch ID. Works with GCS, AWS, and local.
 
     Args:
-        path: the path folder to look in.  Begin with 'gs://' for GCS. Begin with 's3://' for AWS S3.
+        path: The path folder to look in.
+              Begin with 'gs://' for GCS. Begin with 's3://' for AWS S3.
+              Supports wildcards *, **, ?, and [..].
 
-    Returns: the maximum batch id of data in `path`.
+    Returns: The maximum batch id of data in `path`.
+
+    Raises FileNotFoundError if `path` does not exist.
     """
     _LOG.debug(f"Looking for maximum batch id in {path}")
 
-    filesystem = fsspec.open(path).fs
-    if not filesystem.exists(path):
-        _LOG.debug(f"{path} does not exists")
-        return None
+    filesystem = get_filesystem(path)
+    if not filesystem.exists(path.parent):
+        msg = f"{path.parent} does not exists"
+        _LOG.warning(msg)
+        raise FileNotFoundError(msg)
 
-    filenames = get_all_filenames_in_path(path=path)
+    filenames = filesystem.glob(path)
 
-    # just take filename
-    filenames = [filename.split("/")[-1] for filename in filenames]
+    # if there is no files, return 0
+    if len(filenames) == 0:
+        _LOG.debug(f"Did not find any files in {path}")
+        return 0
 
-    # remove suffix
-    filenames = [filename.split(".")[0] for filename in filenames]
+    # just take the stem (the filename without the suffix and without the path)
```

Comment:
This new code should be functionally equivalent, just updated to use `Pathy` functionality.